File: android/src/main/kotlin/com/cloudwebrtc/webrtc/proxy/RtpReceiverProxy.kt
package com.cloudwebrtc.webrtc.proxy

import org.webrtc.RtpReceiver

/**
 * Wrapper around an [RtpReceiver].
 *
 * @param receiver Underlying [RtpReceiver].
 */
class RtpReceiverProxy(receiver: RtpReceiver) : Proxy<RtpReceiver>(receiver) {
  /** [MediaStreamTrackProxy] of this [RtpReceiverProxy]. */
  private var track: MediaStreamTrackProxy = MediaStreamTrackProxy(obj.track()!!)

  init {
    track.replace(obj.track()!!)
    addOnSyncListener { track.replace(obj.track()!!) }
  }

  /** @return Unique ID of the underlying [RtpReceiver]. */
  fun id(): String {
    return obj.id()
  }

  /** @return [MediaStreamTrackProxy] of this [RtpReceiverProxy]. */
  fun getTrack(): MediaStreamTrackProxy {
    return track
  }

  /**
   * Notifies [RtpReceiverProxy] about its [MediaStreamTrackProxy] being removed from the receiver.
   */
  fun notifyRemoved() {
    track.stop()
    track.observableEventBroadcaster().onEnded()
  }
}

File: ios/Classes/proxy/RtpReceiverProxy.swift
import WebRTC

/// Wrapper around an `RTCRtpReceiver`, powering it with additional API.
class RtpReceiverProxy {
  /// Actual underlying [RtpReceiver].
  private var receiver: RTCRtpReceiver

  /// `MediaStreamTrackProxy` of this `RtpReceiverProxy`.
  private var track: MediaStreamTrackProxy

  /// Initializes a new `RtpReceiverProxy` with the provided `RTCRtpReceiver`.
  init(receiver: RTCRtpReceiver) {
    self.receiver = receiver
    self.track = MediaStreamTrackProxy(
      track: self.receiver.track!,
      deviceId: nil,
      source: nil
    )
  }

  /// Returns ID of this `RtpReceiverProxy`.
  func id() -> String {
    self.receiver.receiverId
  }

  /// Returns `MediaStreamTrackProxy` of this `RtpReceiverProxy`.
  func getTrack() -> MediaStreamTrackProxy {
    self.track
  }

  /// Notifies `RtpReceiverProxy` about its `MediaStreamTrackProxy` being
  /// removed from the receiver.
  func notifyRemoved() {
    self.track.notifyEnded()
  }
}
